drivers/dma/dma_pl330.c

```diff
@@ -0,0 +1,610 @@
+/*
+ * Copyright 2020 Broadcom
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <drivers/dma.h>
+#include <init.h>
+#include <string.h>
+#include <soc.h>
+#include <sys/__assert.h>
+#include "dma_pl330.h"
+
+#define LOG_LEVEL CONFIG_DMA_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dma_pl330);
+
+#define DEV_NAME(dev) ((dev)->name)
+#define DEV_CFG(dev) \
+	((const struct dma_pl330_config *const) \
+	 (dev)->config_info)
+#define DEV_DATA(dev) \
+	((struct dma_pl330_dev_data *const)(dev)->driver_data)
+
+#define BYTE_WIDTH(burst_size) (1 << (burst_size))
+
+static int dma_pl330_submit(struct device *dev, uint64_t dst,
+			    uint64_t src, uint32_t channel, uint32_t size);
+
+static void dma_pl330_get_counter(struct dma_pl330_ch_internal *ch_handle,
+				  uint32_t *src_byte_width,
+				  uint32_t *dst_byte_width,
+				  uint32_t *ploop_counter,
+				  uint32_t *presidue)
+{
+	uint32_t srcbytewidth, dstbytewidth;
+	uint32_t loop_counter, residue;
+
+	srcbytewidth = BYTE_WIDTH(ch_handle->src_burst_sz);
+	dstbytewidth = BYTE_WIDTH(ch_handle->dst_burst_sz);
+
+	loop_counter = ch_handle->trans_size /
+		       (srcbytewidth * (ch_handle->src_burst_len + 1));
+
+	if ((loop_counter == 0) && (ch_handle->trans_size > 0)) {
+		residue = ch_handle->trans_size;
+	} else {
+		residue = ch_handle->trans_size - loop_counter *
+			  (srcbytewidth * (ch_handle->src_burst_len + 1));
+	}
+
+	*src_byte_width = srcbytewidth;
+	*dst_byte_width = dstbytewidth;
+	*ploop_counter = loop_counter;
+	*presidue = residue;
+}
+
+static uint32_t dma_pl330_ch_ccr(struct dma_pl330_ch_internal *ch_handle)
+{
+	uint32_t ccr;
+	int secure = ch_handle->nonsec_mode ? SRC_PRI_NONSEC_VALUE :
+		     SRC_PRI_SEC_VALUE;
+
+	ccr = ((ch_handle->dst_cache_ctrl & CC_SRCCCTRL_MASK) <<
+		CC_DSTCCTRL_SHIFT) +
+	       ((ch_handle->nonsec_mode) << CC_DSTNS_SHIFT) +
+	       (ch_handle->dst_burst_len << CC_DSTBRSTLEN_SHIFT) +
+	       (ch_handle->dst_burst_sz << CC_DSTBRSTSIZE_SHIFT) +
+	       (ch_handle->dst_inc << CC_DSTINC_SHIFT) +
+	       ((ch_handle->src_cache_ctrl & CC_SRCCCTRL_MASK) <<
+		CC_SRCCCTRL_SHIFT) +
+	       (secure << CC_SRCPRI_SHIFT) +
+	       (ch_handle->src_burst_len << CC_SRCBRSTLEN_SHIFT)  +
+	       (ch_handle->src_burst_sz << CC_SRCBRSTSIZE_SHIFT)  +
+	       (ch_handle->src_inc << CC_SRCINC_SHIFT);
+
+	return ccr;
+}
+
+static void dma_pl330_calc_burstsz_len(struct dma_pl330_ch_internal *ch_handle,
+				       uintptr_t dst,
+				       uintptr_t src, uint32_t size)
+{
+	int burst_sz, burst_len;
+	uint32_t byte_width;
+
+	burst_sz = MAX_BURST_SIZE_LOG2;
+	/* src, dst and size should be aligned to burst size in bytes */
+	while ((src | dst | size) & ((BYTE_WIDTH(burst_sz)) - 1))
```

Comment:
{ }, we enforce brackets even for 1 statement if/for/while..

Reply:
ack.